File: android/src/main/kotlin/com/example/media_manager/MediaManagerPlugin.kt
package com.example.media_manager

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Environment
import android.util.LruCache
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import java.io.File
import java.util.concurrent.Executors
import io.flutter.plugin.common.PluginRegistry

/** MediaManagerPlugin */
class MediaManagerPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private var activityBinding: ActivityPluginBinding? = null
  private val imageCache = LruCache<String, Bitmap>(20 * 1024 * 1024) // 20MB cache
  private val dispatcher = Executors.newFixedThreadPool(4).asCoroutineDispatcher()
  private val scope = CoroutineScope(dispatcher)

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "media_manager")
    context = flutterPluginBinding.applicationContext
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      "getDirectories" -> {
        getDirectories(result)
      }
      "getDirectoryContents" -> {
        val path = call.argument<String>("path") ?: Environment.getExternalStorageDirectory().path
        getDirectoryContents(path, result)
      }
      "getImagePreview" -> {
        val imagePath = call.argument<String>("path")
        if (imagePath != null) {
          getImagePreview(imagePath, result)
        } else {
          result.error("INVALID_PATH", "Invalid image path", null)
        }
      }
      "clearImageCache" -> {
        clearImageCache()
        result.success(true)
      }
      "requestStoragePermission" -> {
        requestStoragePermission(result)
      }
      "getAllImages" -> {
        getAllFilesByType(result, listOf("jpg", "jpeg", "png", "gif", "bmp", "webp", "tiff", "svg", "ico", "heif", "avif"))
      }
      "getAllVideos" -> {
        getAllFilesByType(result, listOf("mp4", "mov", "m4v", "avi", "mkv", "wmv", "flv", "webm"))
      }
      "getAllAudio" -> {
        getAllFilesByType(result, listOf("mp3", "wav", "m4a", "ogg", "flac", "aac", "wma", "opus"))
      }
      "getAllDocuments" -> {
        getAllFilesByType(result, listOf(// Spreadsheet formats
          "xls", "xlsx", "xlsm", "xlsb", "xlt", "xltx", "xltm",
          "ods", "ots", "csv",

          // Presentation formats
          "ppt", "pptx", "pptm", "pps", "ppsx", "ppsm",
          "pot", "potx", "potm", "odp", "otp",

          // Programming/Code files
          "dart", "php", "js", "jsx", "ts", "tsx", "py", "java",
          "kt", "kts", "cpp", "c", "h", "hpp", "cs", "go", "rb",
          "swift", "m", "mm", "sh", "bash", "ps1", "bat", "cmd",
          "pl", "pm", "lua", "sql", "json", "yaml", "yml", "toml",
          "ini", "cfg", "conf", "gradle", "properties", "asm",
          "s", "v", "vhdl", "verilog", "r", "d", "f", "f90",
          "coffee", "scala", "groovy", "clj", "cljc", "cljs",
          "edn", "ex", "exs", "elm", "erl", "hrl", "fs", "fsx",
          "fsi", "ml", "mli", "nim", "pde", "pp", "pas", "lisp",
          "cl", "scm", "ss", "rkt", "st", "tcl", "vhd", "vhdl",

          // Web development
          "vue", "svelte", "astro", "php", "phtml", "twig",
          "mustache", "hbs", "ejs", "haml", "scss", "sass",
          "less", "styl", "stylus", "coffee", "litcoffee",
          "graphql", "gql", "wasm", "wat",

          // CAD/Design files
          "dwg", "dxf", "dgn", "stl", "obj", "3ds", "max", "blend",
          "skp", "ai", "psd", "eps", "svg", "fig", "xd", "indd",
          "cdr", "afdesign", "afphoto", "afpub", "sketch", "ase",
          "aseprite", "clip", "csp", "kra", "pdn", "procreate",
          "xcf", "pd", "ps", "psb",

          // Archive formats
          "zip", "rar", "7z", "tar", "gz", "bz2", "xz", "lz",
          "lzma", "lzh", "arj", "cab", "iso", "img", "dmg",
          "pkg", "deb", "rpm", "msi", "crx",

          // Ebook formats
          "djvu", "chm", "oxps", "xps", "cbr", "cbz", "cb7",
          "cbt", "cba", "ibooks",

          // Database files
          "db", "sqlite", "sqlite3", "mdb", "accdb", "frm", "myd",
          "myi", "ibd", "mdf", "ldf", "sdf", "nsf", "kdbx",
          "gdb", "fp7", "neo", "db3",

          // Other common document formats
          "md", "markdown", "tex", "log", "pages", "wpd", "wps",
          "abw", "zabw", "123", "602", "wk1", "wk3", "wk4", "wq1",
          "wq2", "xlw", "pmd", "sxw", "stw", "vor", "sxg", "otg",
          "odg", "odc", "odf", "odi", "oxt", "sxc", "stc", "sxd",
          "std", "sxi", "sti", "sxm", "mml", "smf", "odfl", "hwp",
          "hwt", "cell", "numbers", "key", "numbers-tef", "key-tef",
          "papers", "uof", "uot", "uos", "uop", "wmf", "emf", "cgm",
          "vsd", "vsdx", "vss", "vst", "vdx", "vsdm", "vssm", "vstm",
          "pub", "xpub", "sldprt", "sldasm", "slddrw", "prt", "asm",
          "drw", "f3d", "f3z", "iam", "ipt", "catproduct", "catpart",
          "catdrawing", "cgr", "dlv", "exp", "model", "par", "psm",
          "pwd", "session", "sim", "sldlfp", "sldlfp", "std", "stl",
          "stp", "unv", "xas", "xpr", "3dm", "3mf", "amf", "dae",
          "fbx", "glb", "gltf", "iges", "igs", "jt", "obj", "ply",
          "stp", "step", "vrml", "wrl", "x3d", "x3db", "x3dv", "xgl",
          "zgl", "bim", "ifc", "dwf", "dwfx", "nwd", "nwf", "nwc",
          "pln", "pla", "pod", "skb", "layout", "template", "lcd",
          "max", "ma", "mb", "hip", "hiplc", "hda", "hdanc", "usd",
          "usda", "usdc", "usdz", "abc", "vdb", "bgeo", "geo", "pc",
          "pdc", "pdb", "ptc", "rib", "rs", "sc", "scn", "v", "veg",
          "vfb", "vfz", "vob", "vpe", "vray", "vrmesh", "vscene",
          "vue", "w3d", "wings", "wrl", "x", "x3d", "x3dv", "xsi",
          "zbrush"))
      }
      "getAllZipFiles" -> {
        getAllFilesByType(result, listOf("zip", "rar", "7z", "tar", "gz", "bz2", "xz"))
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun getDirectories(result: Result) {
    scope.launch {
      try {
        val directories = mutableListOf<Map<String, Any>>()
        val externalStorage = Environment.getExternalStorageDirectory()

        externalStorage.listFiles()?.forEach { file ->
          if (file.isDirectory) {
            val dirInfo = mutableMapOf<String, Any>()
            dirInfo["name"] = file.name
            dirInfo["path"] = file.absolutePath
            directories.add(dirInfo)
          }
        }

        // Sort directories alphabetically
        directories.sortWith(compareBy { it["name"] as String })

        // Convert to a format that Flutter can handle
        val flutterDirectories = directories.map { dir ->
          mapOf(
            "name" to (dir["name"] as String),
            "path" to (dir["path"] as String)
          )
        }

        withContext(Dispatchers.Main) {
          result.success(flutterDirectories)
        }
      } catch (e: Exception) {
        withContext(Dispatchers.Main) {
          result.error("DIRECTORY_ACCESS_ERROR", "Error accessing directories: ${e.message}", null)
        }
      }
    }
  }

  private fun getDirectoryContents(directoryPath: String, result: Result) {
    scope.launch {
      try {
        val directory = File(directoryPath)
        if (!directory.exists() || !directory.isDirectory) {
          withContext(Dispatchers.Main) {
            result.error("INVALID_DIRECTORY", "Invalid directory path: $directoryPath", null)
          }
          return@launch
        }

        val contents = mutableListOf<Map<String, Any>>()
        directory.listFiles()?.forEach { file ->
          val fileInfo = mutableMapOf<String, Any>()
          fileInfo["name"] = file.name
          fileInfo["path"] = file.absolutePath
          fileInfo["isDirectory"] = file.isDirectory
          fileInfo["size"] = file.length()
          fileInfo["lastModified"] = file.lastModified()

          // Add file type and additional metadata
          if (!file.isDirectory) {
            val extension = file.extension.lowercase()
            fileInfo["type"] = when {
              listOf("jpg", "jpeg", "png", "gif", "bmp", "webp").contains(extension) -> "image"
              listOf("mp4", "avi", "mov", "mkv", "wmv", "flv").contains(extension) -> "video"
              listOf("mp3", "wav", "ogg", "m4a", "flac").contains(extension) -> "audio"
              listOf("pdf", "doc", "docx", "txt", "rtf").contains(extension) -> "document"
              listOf("zip", "rar", "7z").contains(extension) -> "zip"
              else -> "other"
            }

            // Add file extension
            fileInfo["extension"] = extension

            // Add readable file size
            fileInfo["readableSize"] = formatFileSize(file.length())
          } else {
            fileInfo["type"] = "directory"
            fileInfo["extension"] = ""
            fileInfo["readableSize"] = ""
          }

          contents.add(fileInfo)
        }

        // Sort contents: directories first, then files alphabetically
        contents.sortWith(compareBy(
          { (it["isDirectory"] as Boolean).not() },
          { it["name"] as String }
        ))

        // Convert to a format that Flutter can handle
        val flutterContents = contents.map { item ->
          mapOf(
            "name" to (item["name"] as String),
            "path" to (item["path"] as String),
            "isDirectory" to (item["isDirectory"] as Boolean),
            "type" to (item["type"] as String),
            "extension" to (item["extension"] as String),
            "readableSize" to (item["readableSize"] as String)
          )
        }

        withContext(Dispatchers.Main) {
          result.success(flutterContents)
        }
      } catch (e: Exception) {
        withContext(Dispatchers.Main) {
          result.error("FILE_ACCESS_ERROR", "Error accessing files: ${e.message}", null)
        }
      }
    }
  }

  private fun getImagePreview(imagePath: String, result: Result) {
    scope.launch {
      try {
        // Check cache
        val cachedBitmap = imageCache.get(imagePath)
        if (cachedBitmap != null) {
          val byteArray = compressBitmapToByteArray(cachedBitmap)
          withContext(Dispatchers.Main) {
            result.success(byteArray)
          }
          return@launch
        }

        // Load and cache image
        val file = File(imagePath)
        if (!file.exists() || !file.isFile) {
          withContext(Dispatchers.Main) {
            result.error("INVALID_IMAGE", "Invalid image", null)
          }
          return@launch
        }

        // Calculate optimal sample size based on target dimensions
        val options = BitmapFactory.Options().apply {
          inJustDecodeBounds = true
          BitmapFactory.decodeFile(imagePath, this)
          val targetWidth = 800  // Increased from previous value
          val targetHeight = 800  // Increased from previous value
          val scaleFactor = minOf(
            outWidth / targetWidth,
            outHeight / targetHeight
          ).coerceAtLeast(1)
          inSampleSize = scaleFactor
          inJustDecodeBounds = false
        }

        val bitmap = BitmapFactory.decodeFile(imagePath, options)
        if (bitmap != null) {
          // Save to cache
          imageCache.put(imagePath, bitmap)
          val byteArray = compressBitmapToByteArray(bitmap)
          withContext(Dispatchers.Main) {
            result.success(byteArray)
          }
        } else {
          withContext(Dispatchers.Main) {
            result.error("IMAGE_DECODE_ERROR", "Error decoding image", null)
          }
        }
      } catch (e: Exception) {
        withContext(Dispatchers.Main) {
          result.error("IMAGE_PREVIEW_ERROR", "Error generating preview: ${e.message}", null)
        }
      }
    }
  }

  private fun compressBitmapToByteArray(bitmap: Bitmap): ByteArray {
    return bitmap.let {
      val outputStream = java.io.ByteArrayOutputStream()
      // Increased quality from 80 to 90
      it.compress(Bitmap.CompressFormat.JPEG, 90, outputStream)
      outputStream.toByteArray()
    }
  }

  private fun clearImageCache() {
    imageCache.evictAll()
  }

  private fun requestStoragePermission(result: Result) {
    val permissions = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
      arrayOf(
        Manifest.permission.READ_MEDIA_IMAGES,
        Manifest.permission.READ_MEDIA_VIDEO,
        Manifest.permission.READ_MEDIA_AUDIO
      )
    } else {
      arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
    }

    val hasAllPermissions = permissions.all {
      ContextCompat.checkSelfPermission(context, it) == PackageManager.PERMISSION_GRANTED
    }

    if (hasAllPermissions) {
      result.success(true)
    } else {
      activityBinding?.let { binding ->
        val listener = object : PluginRegistry.RequestPermissionsResultListener {
          override fun onRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<String>,
            grantResults: IntArray
          ): Boolean {
            if (requestCode == 1) {
              val granted = grantResults.isNotEmpty() && 
                          grantResults.all { it == PackageManager.PERMISSION_GRANTED }
              result.success(granted)
              binding.removeRequestPermissionsResultListener(this)
              return true
            }
            return false
          }
        }
        
        binding.addRequestPermissionsResultListener(listener)
        ActivityCompat.requestPermissions(
          binding.activity,
          permissions,
          1
        )
      } ?: run {
        result.error("ACTIVITY_NOT_AVAILABLE", "Activity is not available for permission request", null)
      }
    }
  }

  private fun getAllFilesByType(result: Result, extensions: List<String>) {
    scope.launch {
      try {
        val files = mutableListOf<String>()

        fun scanDirectoryForFiles(directory: File) {
          directory.listFiles()?.forEach { file ->
            if (file.isDirectory) {
              scanDirectoryForFiles(file)
            } else if (extensions.contains(file.extension.lowercase())) {
              files.add(file.absolutePath)
            }
          }
        }

        val externalStorage = Environment.getExternalStorageDirectory()
        scanDirectoryForFiles(externalStorage)

        withContext(Dispatchers.Main) {
          result.success(files)
        }
      } catch (e: Exception) {
        withContext(Dispatchers.Main) {
          result.error("FILE_SCAN_ERROR", "Error scanning files: ${e.message}", null)
        }
      }
    }
  }

  private fun formatFileSize(size: Long): String {
    val units = arrayOf("B", "KB", "MB", "GB", "TB")
    var fileSize = size.toDouble()
    var unitIndex = 0

    while (fileSize >= 1024 && unitIndex < units.size - 1) {
      fileSize /= 1024
      unitIndex++
    }

    return "%.2f %s".format(fileSize, units[unitIndex])
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    scope.cancel()
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activityBinding = binding
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activityBinding = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activityBinding = binding
  }

  override fun onDetachedFromActivity() {
    activityBinding = null
  }
}

File: ios/Classes/MediaManagerPlugin.swift
import Flutter
import UIKit
import Photos

public class MediaManagerPlugin: NSObject, FlutterPlugin {
    private var imageCache = NSCache<NSString, UIImage>()
    private var permissionHandler: PermissionHandler?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "media_manager", binaryMessenger: registrar.messenger())
        let instance = MediaManagerPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        case "getDirectories":
            getDirectories(result: result)

        case "getDirectoryContents":
            if let args = call.arguments as? [String: Any],
               let path = args["path"] as? String {
                getDirectoryContents(path: path, result: result)
            } else {
                result(FlutterError(code: "INVALID_PATH",
                                  message: "Invalid directory path",
                                  details: nil))
            }

        case "getImagePreview":
            if let args = call.arguments as? [String: Any],
               let path = args["path"] as? String {
                getImagePreview(path: path, result: result)
            } else {
                result(FlutterError(code: "INVALID_PATH",
                                  message: "Invalid image path",
                                  details: nil))
            }

        case "clearImageCache":
            clearImageCache()
            result(true)

        case "requestStoragePermission":
            requestStoragePermission(result: result)

        case "getAllImages":
            getAllFilesByType(result: result, extensions: ["jpg", "jpeg", "png", "gif", "bmp", "webp", "tiff", "svg", "ico", "heif", "avif"])

        case "getAllVideos":
            getAllFilesByType(result: result, extensions: ["mp4", "mov", "m4v", "avi", "mkv", "wmv", "flv", "webm"])

        case "getAllAudio":
            getAllFilesByType(result: result, extensions: ["mp3", "wav", "m4a", "ogg", "flac", "aac", "wma", "opus"])

        case "getAllDocuments":
            getAllFilesByType(
                result: result,
                extensions: [
                    // Document formats
                    "pdf", "doc", "docx", "docm", "dot", "dotx", "dotm",
                    "txt", "rtf", "odt", "ott", "odm", "oth",
                    "xml", "html", "htm", "xhtml", "mhtml",
                    "epub", "mobi", "azw", "fb2",
                    
                    // Spreadsheet formats
                    "xls", "xlsx", "xlsm", "xlsb", "xlt", "xltx", "xltm",
                    "ods", "ots", "csv",
                    
                    // Presentation formats
                    "ppt", "pptx", "pptm", "pps", "ppsx", "ppsm",
                    "pot", "potx", "potm", "odp", "otp",
                    
                    // Programming/Code files
                    "dart", "php", "js", "jsx", "ts", "tsx", "py", "java",
                    "kt", "kts", "cpp", "c", "h", "hpp", "cs", "go", "rb",
                    "swift", "m", "mm", "sh", "bash", "ps1", "bat", "cmd",
                    "pl", "pm", "lua", "sql", "json", "yaml", "yml", "toml",
                    "ini", "cfg", "conf", "gradle", "properties", "asm",
                    "s", "v", "vhdl", "verilog", "r", "d", "f", "f90",
                    "coffee", "scala", "groovy", "clj", "cljc", "cljs",
                    "edn", "ex", "exs", "elm", "erl", "hrl", "fs", "fsx",
                    "fsi", "ml", "mli", "nim", "pde", "pp", "pas", "lisp",
                    "cl", "scm", "ss", "rkt", "st", "tcl", "vhd", "vhdl",
                    
                    // Web development
                    "vue", "svelte", "astro", "php", "phtml", "twig",
                    "mustache", "hbs", "ejs", "haml", "scss", "sass",
                    "less", "styl", "stylus", "coffee", "litcoffee",
                    "graphql", "gql", "wasm", "wat",
                    
                    // CAD/Design files
                    "dwg", "dxf", "dgn", "stl", "obj", "3ds", "max", "blend",
                    "skp", "ai", "psd", "eps", "svg", "fig", "xd", "indd",
                    "cdr", "afdesign", "afphoto", "afpub", "sketch", "ase",
                    "aseprite", "clip", "csp", "kra", "pdn", "procreate",
                    "xcf", "pd", "ps", "psb",
                    
                    // Archive formats
                    "zip", "rar", "7z", "tar", "gz", "bz2", "xz", "lz",
                    "lzma", "lzh", "arj", "cab", "iso", "img", "dmg",
                    "pkg", "deb", "rpm", "msi", "crx",
                    
                    // Ebook formats
                    "djvu", "chm", "oxps", "xps", "cbr", "cbz", "cb7",
                    "cbt", "cba", "ibooks",
                    
                    // Database files
                    "db", "sqlite", "sqlite3", "mdb", "accdb", "frm", "myd",
                    "myi", "ibd", "mdf", "ldf", "sdf", "nsf", "kdbx",
                    "gdb", "fp7", "neo", "db3",
                    
                    // Other common document formats
                    "md", "markdown", "tex", "log", "pages", "wpd", "wps",
                    "abw", "zabw", "123", "602", "wk1", "wk3", "wk4", "wq1",
                    "wq2", "xlw", "pmd", "sxw", "stw", "vor", "sxg", "otg",
                    "odg", "odc", "odf", "odi", "oxt", "sxc", "stc", "sxd",
                    "std", "sxi", "sti", "sxm", "mml", "smf", "odfl", "hwp",
                    "hwt", "cell", "numbers", "key", "numbers-tef", "key-tef",
                    "papers", "uof", "uot", "uos", "uop", "wmf", "emf", "cgm",
                    "vsd", "vsdx", "vss", "vst", "vdx", "vsdm", "vssm", "vstm",
                    "pub", "xpub", "sldprt", "sldasm", "slddrw", "prt", "asm",
                    "drw", "f3d", "f3z", "iam", "ipt", "catproduct", "catpart",
                    "catdrawing", "cgr", "dlv", "exp", "model", "par", "psm",
                    "pwd", "session", "sim", "sldlfp", "sldlfp", "std", "stl",
                    "stp", "unv", "xas", "xpr", "3dm", "3mf", "amf", "dae",
                    "fbx", "glb", "gltf", "iges", "igs", "jt", "obj", "ply",
                    "stp", "step", "vrml", "wrl", "x3d", "x3db", "x3dv", "xgl",
                    "zgl", "bim", "ifc", "dwf", "dwfx", "nwd", "nwf", "nwc",
                    "pln", "pla", "pod", "skb", "layout", "template", "lcd",
                    "max", "ma", "mb", "hip", "hiplc", "hda", "hdanc", "usd",
                    "usda", "usdc", "usdz", "abc", "vdb", "bgeo", "geo", "pc",
                    "pdc", "pdb", "ptc", "rib", "rs", "sc", "scn", "v", "veg",
                    "vfb", "vfz", "vob", "vpe", "vray", "vrmesh", "vscene",
                    "vue", "w3d", "wings", "wrl", "x", "x3d", "x3dv", "xsi",
                    "zbrush"
                ]
            )

        case "getAllZipFiles":
            getAllFilesByType(result: result, extensions: ["zip", "rar", "7z", "tar", "gz", "bz2", "xz"])

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func getDirectories(result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async {
            let fileManager = FileManager.default
            let documentsPath = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
            let homePath = documentsPath.deletingLastPathComponent()

            do {
                let contents = try fileManager.contentsOfDirectory(at: homePath,
                                                                  includingPropertiesForKeys: nil,
                                                                  options: [.skipsHiddenFiles])

                let directories = contents.filter { $0.hasDirectoryPath }
                    .map { url -> [String: Any] in
                        return [
                            "name": url.lastPathComponent,
                            "path": url.path
                        ]
                    }
                    .sorted { ($0["name"] as! String) < ($1["name"] as! String) }

                DispatchQueue.main.async {
                    result(directories)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "DIRECTORY_ACCESS_ERROR",
                                      message: error.localizedDescription,
                                      details: nil))
                }
            }
        }
    }

    private func getDirectoryContents(path: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async {
            let fileManager = FileManager.default
            let directoryURL = URL(fileURLWithPath: path)

            do {
                // Only using fileSizeKey to avoid iOS version compatibility issues
                let contents = try fileManager.contentsOfDirectory(at: directoryURL,
                                                                  includingPropertiesForKeys: [.fileSizeKey, .creationDateKey],
                                                                  options: [.skipsHiddenFiles])

                let items = contents.map { url -> [String: Any] in
                    let isDirectory = url.hasDirectoryPath
                    var fileSize: Int64 = 0
                    
                    do {
                        let resourceValues = try url.resourceValues(forKeys: [.fileSizeKey])
                        fileSize = Int64(resourceValues.fileSize ?? 0)
                    } catch {
                        print("Error getting file size: \(error)")
                    }
                    
                    let fileExt = url.pathExtension.lowercased()
                    var lastModified: TimeInterval = 0
                    
                    do {
                        if let date = try url.resourceValues(forKeys: [.creationDateKey]).creationDate {
                            lastModified = date.timeIntervalSince1970
                        }
                    } catch {
                        print("Error getting creation date: \(error)")
                    }

                    return [
                        "name": url.lastPathComponent,
                        "path": url.path,
                        "isDirectory": isDirectory,
                        "type": isDirectory ? "directory" : self.getFileType(fileExt),
                        "extension": fileExt,
                        "size": fileSize,
                        "readableSize": self.formatFileSize(fileSize),
                        "lastModified": lastModified
                    ]
                }
                .sorted { (($0["isDirectory"] as! Bool) && !($1["isDirectory"] as! Bool)) ||
                        (($0["isDirectory"] as! Bool) == ($1["isDirectory"] as! Bool) &&
                         ($0["name"] as! String) < ($1["name"] as! String)) }

                DispatchQueue.main.async {
                    result(items)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "FILE_ACCESS_ERROR",
                                      message: error.localizedDescription,
                                      details: nil))
                }
            }
        }
    }

    private func getImagePreview(path: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async {
            // Check cache first
            if let cachedImage = self.imageCache.object(forKey: path as NSString) {
                if let imageData = cachedImage.jpegData(compressionQuality: 0.9) {
                    DispatchQueue.main.async {
                        result(imageData)
                    }
                    return
                }
            }

            if let image = UIImage(contentsOfFile: path) {
                let resizedImage = self.resizeImage(image, targetSize: CGSize(width: 800, height: 800))
                self.imageCache.setObject(resizedImage, forKey: path as NSString)

                if let imageData = resizedImage.jpegData(compressionQuality: 0.9) {
                    DispatchQueue.main.async {
                        result(imageData)
                    }
                } else {
                    DispatchQueue.main.async {
                        result(FlutterError(code: "IMAGE_COMPRESSION_ERROR",
                                          message: "Failed to compress image",
                                          details: nil))
                    }
                }
            } else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "IMAGE_LOAD_ERROR",
                                      message: "Failed to load image",
                                      details: nil))
                }
            }
        }
    }

    private func clearImageCache() {
        imageCache.removeAllObjects()
    }

    private func requestStoragePermission(result: @escaping FlutterResult) {
        // iOS doesn't have general storage permissions like Android
        // but requires photo library access for accessing photos
        PHPhotoLibrary.requestAuthorization { status in
            DispatchQueue.main.async {
                result(status == .authorized)
            }
        }
    }

    private func getAllFilesByType(result: @escaping FlutterResult, extensions: [String]) {
        DispatchQueue.global(qos: .userInitiated).async {
            let fileManager = FileManager.default
            let documentsPath = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
            let homePath = documentsPath.deletingLastPathComponent()

            var files: [String] = []

            func scanDirectory(_ url: URL) {
                do {
                    let contents = try fileManager.contentsOfDirectory(at: url,
                                                                      includingPropertiesForKeys: nil,
                                                                      options: [.skipsHiddenFiles])

                    for item in contents {
                        if item.hasDirectoryPath {
                            scanDirectory(item)
                        } else if extensions.contains(item.pathExtension.lowercased()) {
                            files.append(item.path)
                        }
                    }
                } catch {
                    print("Error scanning directory: \(error.localizedDescription)")
                }
            }

            scanDirectory(homePath)
            
            DispatchQueue.main.async {
                result(files)
            }
        }
    }

    private func getFileType(_ fileExt: String) -> String {
        switch fileExt {
        case "jpg", "jpeg", "png", "gif", "bmp", "webp", "tiff", "svg", "ico", "heif", "avif":
            return "image"
        case "mp4", "mov", "m4v", "avi", "mkv", "wmv", "flv", "webm":
            return "video"
        case "mp3", "wav", "m4a", "ogg", "flac", "aac", "wma", "opus":
            return "audio"
        case "pdf", "doc", "docx", "txt", "rtf", "md", "markdown":
            return "document"
        case "zip", "rar", "7z", "tar", "gz", "bz2", "xz":
            return "zip"
        default:
            return "other"
        }
    }

    private func formatFileSize(_ size: Int64) -> String {
        let units = ["B", "KB", "MB", "GB", "TB"]
        var fileSize = Double(size)
        var unitIndex = 0

        while fileSize >= 1024 && unitIndex < units.count - 1 {
            fileSize /= 1024
            unitIndex += 1
        }

        return String(format: "%.2f %@", fileSize, units[unitIndex])
    }

    private func resizeImage(_ image: UIImage, targetSize: CGSize) -> UIImage {
        let size = image.size
        let widthRatio = targetSize.width / size.width
        let heightRatio = targetSize.height / size.height
        let ratio = min(widthRatio, heightRatio)

        let newSize = CGSize(width: size.width * ratio, height: size.height * ratio)
        let rect = CGRect(origin: .zero, size: newSize)

        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        image.draw(in: rect)
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return newImage ?? image
    }
}

// Helper class for permission handling (matches Android implementation better)
class PermissionHandler: NSObject {
    var permissionResult: FlutterResult?
    
    func requestPhotoLibraryPermission(result: @escaping FlutterResult) {
        permissionResult = result
        PHPhotoLibrary.requestAuthorization { status in
            DispatchQueue.main.async {
                result(status == .authorized)
            }
        }
    }
}